cleaner.go

```diff
@@ -0,0 +1,362 @@
+		return id
+	}
+	collId, ok := m["id"]
+	if !ok {
+		logger.Warningf("massage failed, no 'id' field: %v", id)
+		return id
+	}
+	return stashId{
+		C:  cStr,
+		Id: collId,
+	}
+}
+
+// NewCollectionCleaner creates an object that can remove transaction tokens from document queues when the transactions
+// have been marked as completed.
+func NewCollectionCleaner(config CollectionConfig) *collectionCleaner {
+	if config.NumBatchTokens == 0 {
+		config.NumBatchTokens = queueBatchSize
+	}
+	if config.MaxRemoveQueue == 0 {
+		config.MaxRemoveQueue = maxMemoryTokens
+	}
+	return &collectionCleaner{
+		config:         config,
+		docIdsToRemove: make([]interface{}, 0),
+		docsToProcess:  make([]txnDocument, 0),
+		tokensToLookup: make([]string, 0),
+		removeIfEmpty:  false,
+		massageId:      nil,
+	}
+}
+
+// NewStashCleaner returns an object suitable for cleaning up all of the txns.stash collection.
+// It is unique in that the document Ids in this collection are special, and if we find all references from a document
+// have been removed, we no longer need the document to exist.
+func NewStashCleaner(config CollectionConfig) *collectionCleaner {
+	return &collectionCleaner{
+		config:         config,
+		docIdsToRemove: make([]interface{}, 0),
+		docsToProcess:  make([]txnDocument, 0),
+		tokensToLookup: make([]string, 0),
+		removeIfEmpty:  true,
+		massageId:      massageStashId,
+	}
+}
+
+// includeDoc queues this doc to be processed. It returns 'true' if the docs should be processed.
+func (cleaner *collectionCleaner) includeDoc(doc txnDocument) (bool, error) {
+	cleaner.stats.DocCount++
+	if cleaner.stats.DocCount%50000 == 0 {
+		maxCount, _ := cleaner.config.Source.Count()
+		logger.Debugf("processed %d/%d docs from %q (removed %d)",
+			cleaner.stats.DocCount, maxCount, cleaner.config.Source.Name, cleaner.stats.RemovedCount)
+	}
+	cleaner.docsToProcess = append(cleaner.docsToProcess, doc)
+	for _, token := range doc.Queue {
+		cleaner.tokensToLookup = append(cleaner.tokensToLookup, token)
+	}
+	return len(cleaner.tokensToLookup) >= cleaner.config.NumBatchTokens, nil
+}
+
+// findCompletedTokens looks at the list of tokens and finds what txns are referenced as completed,
+// and then returns the set of tokens that are completed.
+func (cleaner *collectionCleaner) findCompletedTokens() (map[string]bool, error) {
+	objectIds := make([]bson.ObjectId, 0, len(cleaner.tokensToLookup))
+	// The nonce is generated during preparing, and if 2 flushers race, only one nonce makes it into the final
+	// transaction. However, other nonces can also be considered 'completed'. (afaict, they are ignored, thus won't
+	// be applied and can be considered completed.)
+	for _, token := range cleaner.tokensToLookup {
+		objId := txnTokenToId(token)
+		objectIds = append(objectIds, objId)
+	}
+	query := cleaner.config.Txns.Find(bson.M{"_id": bson.M{"$in": objectIds}, "s": bson.M{"$in": []int{tapplied, taborted}}})
```

Comment:
Idea: given that the txn pruning logic does this query too (for all completed txns), writing the results to a temporary collection, would it make sense to extract this out and use the temporary collection in both places?

This would require a small amount of reorg so that the top level pruning func owns the job of cleaning up the temporary collection. 

It might be good to have a type which wraps the temporary completed txns collection for querying and cleanup. It could be created at the top level and passed in to both the cleaning and pruning steps.